File: kafka/kafka-annotation-processor/src/main/java/ru/tinkoff/kora/kafka/annotation/processor/utils/KafkaUtils.java
package ru.tinkoff.kora.kafka.annotation.processor.utils;

import com.palantir.javapoet.ClassName;
import com.palantir.javapoet.TypeName;
import org.jspecify.annotations.Nullable;
import ru.tinkoff.kora.annotation.processor.common.AnnotationUtils;
import ru.tinkoff.kora.kafka.annotation.processor.KafkaClassNames;

import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;

import static ru.tinkoff.kora.annotation.processor.common.CommonUtils.capitalize;
import static ru.tinkoff.kora.annotation.processor.common.CommonUtils.decapitalize;

public final class KafkaUtils {

    private KafkaUtils() {}

    public static String prepareConsumerTagName(ExecutableElement method) {
        var controllerName = method.getEnclosingElement().getSimpleName().toString();
        var methodName = method.getSimpleName().toString();
        return capitalize(controllerName) + capitalize(methodName) + "Tag";
    }

    public static ClassName prepareConsumerTag(Elements elements, ExecutableElement method) {
        String moduleName = method.getEnclosingElement().getSimpleName().toString() + "Module";
        return ClassName.get(elements.getPackageOf(method).getQualifiedName().toString(), moduleName, prepareConsumerTagName(method));
    }

    @Nullable
    public static ClassName findConsumerUserTag(ExecutableElement method) {
        var annotation = AnnotationUtils.findAnnotation(method, KafkaClassNames.kafkaListener);
        var tag = AnnotationUtils.<TypeMirror>parseAnnotationValueWithoutDefault(annotation, "tag");
        if (tag == null) {
            return null;
        } else {
            return (ClassName) TypeName.get(tag);
        }
    }

    @Nullable
    public static ClassName getConsumerTag(Elements elements, ExecutableElement method) {
        var tags = findConsumerUserTag(method);
        if (tags == null) {
            return prepareConsumerTag(elements, method);
        } else {
            return tags;
        }
    }

    public static String prepareMethodName(ExecutableElement method, String suffix) {
        var controllerName = method.getEnclosingElement().getSimpleName().toString();
        var methodName = method.getSimpleName().toString();
        return decapitalize(controllerName) + capitalize(methodName) + suffix;
    }

    public static boolean isConsumerRecord(TypeMirror tm) {
        return tm instanceof DeclaredType dt && ClassName.get((TypeElement) dt.asElement()).equals(KafkaClassNames.consumerRecord);
    }

    public static boolean isConsumerRecords(TypeMirror tm) {
        return tm instanceof DeclaredType dt && ClassName.get((TypeElement) dt.asElement()).equals(KafkaClassNames.consumerRecords);
    }

    public static boolean isProducerRecord(TypeMirror tm) {
        return tm instanceof DeclaredType dt && ClassName.get((TypeElement) dt.asElement()).equals(KafkaClassNames.producerRecord);
    }

    public static boolean isProducerCallback(TypeMirror tm) {
        return tm instanceof DeclaredType dt && ClassName.get((TypeElement) dt.asElement()).equals(KafkaClassNames.producerCallback);
    }

    public static boolean isHeaders(TypeMirror tm) {
        return tm instanceof DeclaredType dt && ClassName.get((TypeElement) dt.asElement()).equals(KafkaClassNames.headers);
    }

    public static boolean isHeader(TypeMirror tm) {
        return tm instanceof DeclaredType dt && ClassName.get((TypeElement) dt.asElement()).equals(KafkaClassNames.header);
    }

    public static boolean isKeyDeserializationException(TypeMirror tm) {
        return tm instanceof DeclaredType dt && ClassName.get((TypeElement) dt.asElement()).equals(KafkaClassNames.recordKeyDeserializationException);
    }

    public static boolean isValueDeserializationException(TypeMirror tm) {
        return tm instanceof DeclaredType dt && ClassName.get((TypeElement) dt.asElement()).equals(KafkaClassNames.recordValueDeserializationException);
    }

    public static boolean isAnyException(TypeMirror tm) {
        return tm.toString().equals("java.lang.Throwable") || tm.toString().equals("java.lang.Exception");
    }

    public static boolean isConsumer(TypeMirror tm) {
        return tm instanceof DeclaredType dt && ClassName.get((TypeElement) dt.asElement()).equals(KafkaClassNames.consumer);
    }

}

File: kafka/kafka-symbol-processor/src/main/kotlin/ru/tinkoff/kora/kafka/symbol/processor/KafkaUtils.kt
package ru.tinkoff.kora.kafka.symbol.processor

import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.THROWABLE
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.kafka.symbol.processor.KafkaClassNames.consumer
import ru.tinkoff.kora.kafka.symbol.processor.KafkaClassNames.consumerRecord
import ru.tinkoff.kora.kafka.symbol.processor.KafkaClassNames.consumerRecords
import ru.tinkoff.kora.kafka.symbol.processor.KafkaClassNames.recordKeyDeserializationException
import ru.tinkoff.kora.kafka.symbol.processor.KafkaClassNames.recordValueDeserializationException
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findValueNoDefault

object KafkaUtils {
    fun KSFunctionDeclaration.moduleName(suffix: String): String {
        val classDecl = this.parentDeclaration!!
        val prefix = classDecl.simpleName.asString().replaceFirstChar { it.uppercaseChar() }
        val function = this.simpleName.asString().replaceFirstChar { it.uppercaseChar() }

        return "${prefix}${function}${suffix}"
    }

    fun KSFunctionDeclaration.findConsumerUserTag(): TypeName? {
        val listener = findAnnotation(KafkaClassNames.kafkaListener) ?: return null
        val tag = listener.findValueNoDefault<KSType>("tag")
        if (tag == null) {
            return null
        }

        return tag.toTypeName()
    }

    fun KSFunctionDeclaration.consumerTag(): TypeName {
        val userTags = findConsumerUserTag()
        return userTags ?: tagType()
    }

    fun KSFunctionDeclaration.tagType() = ClassName(packageName.asString(), parentDeclaration!!.simpleName.asString() + "Module", tagTypeName())
    fun KSFunctionDeclaration.tagTypeName() = moduleName("Tag")
    fun KSFunctionDeclaration.containerFunName() = moduleName("Container").replaceFirstChar { it.lowercaseChar() }
    fun KSFunctionDeclaration.handlerFunName() = moduleName("Handler").replaceFirstChar { it.lowercaseChar() }
    fun KSFunctionDeclaration.configFunName() = moduleName("Config").replaceFirstChar { it.lowercaseChar() }

    fun KSType.isConsumerRecord() = declaration.let { it is KSClassDeclaration && it.toClassName() == consumerRecord }

    fun KSType.isConsumerRecords() = declaration.let { it is KSClassDeclaration && it.toClassName() == consumerRecords }

    fun KSType.isKeyDeserializationException() = declaration.let { it is KSClassDeclaration && it.toClassName() == recordKeyDeserializationException }

    fun KSType.isValueDeserializationException() = declaration.let { it is KSClassDeclaration && it.toClassName() == recordValueDeserializationException }

    fun KSType.isAnyException() = toTypeName().copy(false).let {
        it is ClassName && (it == THROWABLE || it.toString() == "kotlin.Exception" || it.toString() == "java.lang.Exception" || it.toString() == "java.lang.Throwable")
    }

    fun KSType.isConsumer() = declaration.let { it is KSClassDeclaration && it.toClassName() == consumer }
}


